src/main/java/net/dv8tion/jda/api/utils/FileProxy.java

```diff
@@ -0,0 +1,330 @@
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+/**
+ * A utility class to download files
+ */
+public class FileProxy
+{
+    private static volatile OkHttpClient defaultHttpClient;
+
+    private final String url;
+    private OkHttpClient customHttpClient;
+
+    /**
+     * Constructs a new {@link FileProxy} for the provided URL
+     *
+     * @param url The URL to download from
+     */
+    public FileProxy(@Nonnull String url)
+    {
+        Checks.notNull(url, "URL");
+
+        this.url = url;
+    }
+
+    /**
+     * Sets the default OkHttpClient used by {@link FileProxy} and {@link ImageProxy}
+     * <br>This can still be overridden on a per-instance basis with {@link #withClient(OkHttpClient)}
+     *
+     * @param  httpClient
+     *         The default {@link OkHttpClient} to use while making HTTP requests
+     *
+     * @throws IllegalArgumentException
+     *         If the provided {@link OkHttpClient} is null
+     */
+    public static void setHttpClient(@Nonnull OkHttpClient httpClient)
+    {
+        Checks.notNull(httpClient, "Default OkHttpClient");
+
+        FileProxy.defaultHttpClient = httpClient;
+    }
+
+    /**
+     * Returns the URL that has been passed to this proxy
+     * <br>This URL is always from Discord
+     *
+     * @return The URL of the file
+     */
+    @Nonnull
+    public String getUrl()
+    {
+        return url;
+    }
+
+    /**
+     * Sets the custom OkHttpClient used by this instance, regardless of if {@link #setHttpClient(OkHttpClient)} has been used or not
+     *
+     * @param  customHttpClient
+     *         The custom {@link OkHttpClient} to use while making HTTP requests
+     *
+     * @throws IllegalArgumentException
+     *         If the provided {@link OkHttpClient} is null
+     *
+     * @return This proxy for chaining convenience
+     */
+    @Nonnull
+    public FileProxy withClient(@Nonnull OkHttpClient customHttpClient)
+    {
+        Checks.notNull(customHttpClient, "Custom HTTP client");
+
+        this.customHttpClient = customHttpClient;
+
+        return this;
+    }
+
+
+    // INTERNAL DOWNLOAD METHODS
+
+    protected OkHttpClient getHttpClient() {
+        // Return custom HTTP client if set
+        if (customHttpClient != null)
+        {
+            return customHttpClient;
+        }
+
+        // Otherwise, see if a default one has been assigned
+        //  If there is no client then create a default one
+        if (defaultHttpClient == null)
+        {
+            synchronized (this) {
+                if (defaultHttpClient == null)
+                {
+                    defaultHttpClient = new OkHttpClient();
+                }
+            }
+        }
```

Comment:
```suggestion
        if (customHttpClient != null)
            return customHttpClient;

        // Otherwise, see if a default one has been assigned
        //  If there is no client then create a default one
        if (defaultHttpClient == null)
        {
            synchronized (this)
            {
                if (defaultHttpClient == null)
                    defaultHttpClient = new OkHttpClient();
            }
        }
```